File: downloader-api/src/extractor.rs
use std::process::Stdio;
use tokio::process::Command;
use crate::models::{ExtractResponse, VideoFormat};

/// Extract video information using yt-dlp
pub async fn extract_video_info(url: &str) -> Result<ExtractResponse, String> {
    // Run yt-dlp with JSON output
    let output = Command::new("yt-dlp")
        .args([
            "--dump-json",
            "--no-warnings",
            "--no-playlist",
            url,
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .await
        .map_err(|e| format!("Failed to run yt-dlp: {}. Is yt-dlp installed?", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("yt-dlp error: {}", stderr.trim()));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let json: serde_json::Value = serde_json::from_str(&stdout)
        .map_err(|e| format!("Failed to parse yt-dlp output: {}", e))?;

    // Extract platform
    let platform = detect_platform(url);

    // Extract title
    let title = json["title"]
        .as_str()
        .unwrap_or("Untitled")
        .to_string();

    // Extract thumbnail
    let thumbnail = json["thumbnail"].as_str().map(String::from);

    // Extract formats
    let formats = extract_formats(&json);

    Ok(ExtractResponse::new(platform, title, thumbnail, formats))
}

/// Detect platform from URL
fn detect_platform(url: &str) -> String {
    let url_lower = url.to_lowercase();
    if url_lower.contains("instagram.com") {
        "instagram".to_string()
    } else if url_lower.contains("tiktok.com") {
        "tiktok".to_string()
    } else if url_lower.contains("twitter.com") || url_lower.contains("x.com") {
        "twitter".to_string()
    } else {
        "unknown".to_string()
    }
}

/// Extract video formats from yt-dlp JSON output
fn extract_formats(json: &serde_json::Value) -> Vec<VideoFormat> {
    let mut formats = Vec::new();

    // Try to get the best format directly
    if let Some(url) = json["url"].as_str() {
        formats.push(VideoFormat {
            quality: "best".to_string(),
            url: url.to_string(),
            ext: json["ext"].as_str().unwrap_or("mp4").to_string(),
            filesize: json["filesize"].as_u64(),
        });
        return formats;
    }

    // Parse formats array
    if let Some(format_list) = json["formats"].as_array() {
        // Filter video formats and sort by quality
        let mut video_formats: Vec<_> = format_list
            .iter()
            .filter(|f| {
                let vcodec = f["vcodec"].as_str().unwrap_or("none");
                vcodec != "none" && f["url"].as_str().is_some()
            })
            .collect();

        // Sort by height (quality) descending
        video_formats.sort_by(|a, b| {
            let h_a = a["height"].as_u64().unwrap_or(0);
            let h_b = b["height"].as_u64().unwrap_or(0);
            h_b.cmp(&h_a)
        });

        // Take top 3 qualities
        for f in video_formats.into_iter().take(3) {
            let height = f["height"].as_u64().unwrap_or(0);
            let quality = if height > 0 {
                format!("{}p", height)
            } else {
                f["format_note"].as_str().unwrap_or("unknown").to_string()
            };

            formats.push(VideoFormat {
                quality,
                url: f["url"].as_str().unwrap_or("").to_string(),
                ext: f["ext"].as_str().unwrap_or("mp4").to_string(),
                filesize: f["filesize"].as_u64(),
            });
        }
    }

    // Fallback: try requested_downloads
    if formats.is_empty() {
        if let Some(downloads) = json["requested_downloads"].as_array() {
            for d in downloads {
                if let Some(url) = d["url"].as_str() {
                    formats.push(VideoFormat {
                        quality: "best".to_string(),
                        url: url.to_string(),
                        ext: d["ext"].as_str().unwrap_or("mp4").to_string(),
                        filesize: d["filesize"].as_u64(),
                    });
                }
            }
        }
    }

    formats
}

File: packages/api/src/lib/extractor.ts
import type { ExtractResponse, VideoFormat } from "@snatch/shared";

let ytDlpCommand = "yt-dlp";

export function setYtDlpCommandForTest(command: string): void {
	ytDlpCommand = command;
}

export function resetYtDlpCommandForTest(): void {
	ytDlpCommand = "yt-dlp";
}

export async function extractVideoInfo(url: string): Promise<ExtractResponse> {
	const proc = Bun.spawn([ytDlpCommand, "--dump-json", "--no-warnings", "--no-playlist", url], {
		stdout: "pipe",
		stderr: "pipe",
	});

	const timeout = setTimeout(() => {
		proc.kill();
	}, 30_000);

	const stdoutPromise = new Response(proc.stdout).text();
	const stderrPromise = new Response(proc.stderr).text();

	try {
		const exitCode = await proc.exited;
		const [stdout, stderr] = await Promise.all([stdoutPromise, stderrPromise]);
		clearTimeout(timeout);

		if (exitCode !== 0) {
			const firstLine = stderr.trim().split("\n")[0] || "yt-dlp failed";
			throw new Error(`Extraction failed: ${firstLine}`);
		}

		const json = JSON.parse(stdout);
		return parseYtDlpOutput(json, url);
	} finally {
		clearTimeout(timeout);
	}
}

/**
 * Download video using yt-dlp and stream to response.
 * Includes a 60s read timeout to kill stuck processes.
 */
export function downloadVideoStream(url: string, formatId?: string): ReadableStream<Uint8Array> {
	const formatArg = formatId || "best[ext=mp4]/best";
	const proc = Bun.spawn(
		[ytDlpCommand, "-o", "-", "--no-warnings", "--no-playlist", "-f", formatArg, url],
		{
			stdout: "pipe",
			stderr: "pipe",
		},
	);

	const READ_TIMEOUT_MS = 60_000;
	let timeoutId: ReturnType<typeof setTimeout> | null = null;

	const resetTimeout = () => {
		if (timeoutId) clearTimeout(timeoutId);
		timeoutId = setTimeout(() => {
			proc.kill();
		}, READ_TIMEOUT_MS);
	};

	// biome-ignore lint/suspicious/noExplicitAny: Bun ReadableStreamDefaultReader type mismatch between stream/web and global
	let reader: any = null;
	const stderrPromise = new Response(proc.stderr).text();

	return new ReadableStream({
		start() {
			reader = proc.stdout.getReader();
			resetTimeout();
		},
		async pull(controller) {
			if (!reader) return;
			try {
				const { done, value } = await reader.read();
				if (!done) {
					resetTimeout();
					controller.enqueue(value);
					return;
				}

				if (timeoutId) clearTimeout(timeoutId);
				const [exitCode, stderr] = await Promise.all([proc.exited, stderrPromise]);
				if (exitCode !== 0) {
					const firstLine = stderr.trim().split("\n")[0] || "yt-dlp failed";
					controller.error(new Error(`Download failed: ${firstLine}`));
					return;
				}
				controller.close();
			} catch (error) {
				if (timeoutId) clearTimeout(timeoutId);
				controller.error(error);
			}
		},
		cancel() {
			if (timeoutId) clearTimeout(timeoutId);
			reader?.releaseLock();
			proc.kill();
		},
	});
}

/**
 * Parse yt-dlp JSON output into ExtractResponse
 */
// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON output is dynamic
function parseYtDlpOutput(json: any, url: string): ExtractResponse {
	const platform = detectPlatform(url);
	const title = json.title || "Untitled";
	const thumbnail = json.thumbnail || undefined;
	const formats = extractFormats(json);

	return {
		success: true,
		platform,
		title,
		thumbnail,
		formats,
	};
}

/**
 * Extract formats from yt-dlp JSON.
 *
 * Always prefers the `formats[]` array so the user gets a real quality
 * picker. Top-level `url` is only used as a last resort, because yt-dlp
 * exposes it for every video (it's the auto-selected best format) and
 * short-circuiting on it collapses everything to a single option.
 */
// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON output is dynamic
function extractFormats(json: any): VideoFormat[] {
	if (json.formats?.length) {
		// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON is dynamic
		const videoFormats = json.formats.filter((f: any) => f.vcodec !== "none" && f.url);

		// Dedupe by height: same-resolution variants (e.g. TikTok DASH segments
		// numbered -0 / -1) are noise to the user. Keep the largest filesize.
		// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON is dynamic
		const byHeight = new Map<number, any>();
		for (const f of videoFormats) {
			const h = f.height || 0;
			const existing = byHeight.get(h);
			const size = f.filesize || f.filesize_approx || 0;
			const existingSize = existing ? existing.filesize || existing.filesize_approx || 0 : -1;
			if (!existing || size > existingSize) {
				byHeight.set(h, f);
			}
		}

		const ranked = [...byHeight.values()]
			.sort((a, b) => (b.height || 0) - (a.height || 0))
			.slice(0, 3);

		const formats: VideoFormat[] = ranked.map((f) => {
			const height = f.height || 0;
			return {
				format_id: f.format_id || "unknown",
				quality: height > 0 ? `${height}p` : f.format_note || "unknown",
				url: f.url,
				ext: f.ext || "mp4",
				filesize: f.filesize || f.filesize_approx || undefined,
			};
		});

		if (formats.length > 0) return formats;
	}

	// Fallback: requested_downloads (rare — when yt-dlp resolves a single download)
	if (json.requested_downloads?.length) {
		const formats: VideoFormat[] = json.requested_downloads
			// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON is dynamic
			.filter((d: any) => d.url)
			// biome-ignore lint/suspicious/noExplicitAny: yt-dlp JSON is dynamic
			.map((d: any) => ({
				format_id: d.format_id || "best",
				quality: "best",
				url: d.url,
				ext: d.ext || "mp4",
				filesize: d.filesize || d.filesize_approx || undefined,
			}));
		if (formats.length > 0) return formats;
	}

	// Last resort: top-level url (single quality)
	if (json.url) {
		return [
			{
				format_id: json.format_id || "best",
				quality: "best",
				url: json.url,
				ext: json.ext || "mp4",
				filesize: json.filesize || json.filesize_approx || undefined,
			},
		];
	}

	return [];
}

/**
 * Detect platform from URL
 */
function detectPlatform(url: string): string {
	const host = (() => {
		try {
			return new URL(url).hostname.toLowerCase();
		} catch {
			return "";
		}
	})();

	if (host.includes("tiktok.com")) return "tiktok";
	if (host.includes("x.com") || host.includes("twitter.com")) return "twitter";
	return "unknown";
}
